### PR TITLE
SCC-5470: Index name, title, role, label as display components

### PR DESCRIPTION
**Pull request by 7emansell**
- Continuing to index contributors, creators, and series with their literal values (that we browse/search/filter with) but replacing the `_displayPacked` fields with the new and improved `_displayComponents`
- Relatively straightforward for contributors/creators, since the `buildBrowseTermData`'s `ContributorVarField` model is meant to be used on these MARC fields
- For `seriesUniformTitle` and `seriesAddedEntry`s, we still need to strip the `$v` and `$x` subfields off of the `title` to maintain what we ended up with [last time we tried to improve series](https://docs.google.com/document/d/1jjg6Ns7xDw8q_3B0hWdaOOn_TSiI9uXnZWzP1FQtAZQ/edit?tab=t.0)

So if you want to get into the real weeds:
Series
490 440
prefixes + **name** + title  + role
with **name** linked to /search?filters[series][0]=[**name**]

Series added entries and series uniform title 
800
810
811
830

prefixes + **name + title - $v - $x** $v $x + role 
with **name + title - $v - $x** linked to /search?filters[series][0]=[**name + title - $v - $x**]

Example diff:
```
Bib metadata diff for b16470380:
  {
    "added": {
      "contributor_displayComponents": {
        "locallyBuiltValue": {
          "0": {
            "name": "Spada, Pietro, 1935-",
            "title": "",
            "role": "editor",
            "label": "Spada, Pietro, 1935-, editor"
          }
        }
      },
      "creator_displayComponents": {
        "locallyBuiltValue": {
          "0": {
            "name": "Rossini, Gioacchino, 1792-1868",
            "title": "",
            "role": "",
            "label": "Rossini, Gioacchino, 1792-1868"
          }
        }
      },
      "series_displayComponents": {
        "locallyBuiltValue": {
          "0": {
            "name": "Inediti e rarità rossiniane ;",
            "title": "15",
            "role": "",
            "label": "Inediti e rarità rossiniane ; 15"
          }
        }
      },
      "seriesUniformTitle_displayComponents": {
        "locallyBuiltValue": {
          "0": {
            "name": "Rossini, Gioacchino 1792-1868",
            "title": "Works. Selections (Boccaccini & Spada editore) ;",
            "role": "",
            "label": "Rossini, Gioacchino 1792-1868. Works. Selections (Boccaccini & Spada editore) ; 15"
          }
        }
      }
    },
    "deleted": {
      "contributors_displayPacked": {
        "actualIndexedValue": {
          "0": "Spada, Pietro, 1935-||Spada, Pietro, 1935-, editor"
        }
      },
      "creators_displayPacked": {
        "actualIndexedValue": {
          "0": "Rossini, Gioacchino, 1792-1868||Rossini, Gioacchino, 1792-1868"
        }
      },
      "parallelCreators_displayPacked": {
        "actualIndexedValue": {}
      },
      "parallelContributors_displayPacked": {
        "actualIndexedValue": {}
      },
      "parallelSeriesAddedEntry": {
        "actualIndexedValue": {}
      },
      "parallelSeriesAddedEntry_displayPacked": {
        "actualIndexedValue": {}
      },
      "seriesAddedEntry": {
        "actualIndexedValue": {}
      },
      "series_displayPacked": {
        "actualIndexedValue": {
          "0": "Inediti e rarità rossiniane ;||Inediti e rarità rossiniane ; 15"
        }
      },
      "seriesUniformTitle_displayPacked": {
        "actualIndexedValue": {
          "0": "Rossini, Gioacchino 1792-1868. Works. Selections (Boccaccini & Spada editore) ;||Rossini, Gioacchino 1792-1868. Works. Selections (Boccaccini & Spada editore) ; 15."
        }
      },
      "seriesAddedEntry_displayPacked": {
        "actualIndexedValue": {}
      }
    },
    "updated": {
      "parallelCreatorLiteral": {
        "actualIndexedValue": {},
        "locallyBuiltValue": {}
      },
      "parallelContributorLiteral": {
        "actualIndexedValue": {},
        "locallyBuiltValue": {}
      },
      "seriesUniformTitle": {
        "actualIndexedValue": {
          "0": "Rossini, Gioacchino 1792-1868. Works. Selections (Boccaccini & Spada editore) ;"
        },
        "locallyBuiltValue": {
          "0": "Rossini, Gioacchino 1792-1868 Works. Selections (Boccaccini & Spada editore) ;"
        }
      },
      "updatedAt": {
        "actualIndexedValue": 1784200535660,
        "locallyBuiltValue": 1787233746095
      }
    }
  }

```

### CODE REVIEW

File: lib/es-models/bib.js [hunk 1]
```diff
@@ -815,54 +793,61 @@ class EsBib extends EsBase {
...
+    return this._extractDisplayComponents(terms, { isVariant: true })
+  }
+
+  parallelSeriesUniformTitle_displayComponents () {
+    const terms = this.browseTermData.seriesUniformTitle.filter(term => term && term.variant)
+    return this._extractDisplayComponents(terms, { isVariant: true, stripTrailingVX: true })
+  }
+
+  parallelSeriesAddedEntry_displayComponents () {
+    const terms = this.browseTermData.seriesAddedEntry.filter(term => term && term.variant)
```
Comment (charmingduchess) [inline lib/es-models/bib.js]: why did you remove the null return values? i think we need to maintain the same length of arrays so parallels are returned at the same index as the primary values.

File: lib/es-models/bib.js [hunk 2]
```diff
@@ -815,54 +793,61 @@ class EsBib extends EsBase {
...
+    return this._extractDisplayComponents(terms, { isVariant: true, stripTrailingVX: true })
```
Comment (charmingduchess) [inline lib/es-models/bib.js]: This interface could be cleaned up by providing `_extractParallelDisplayComponents`, `_extractParallelSeriesDisplayComponents`, and `_extractSeriesDisplayComponents`. Up to you, but that's the pattern we ended up with for `valueToIndexForBasicMapping`/`parallelValueToIndex`.

File: lib/es-models/bib.js [hunk 3]
```diff
@@ -815,54 +793,61 @@ class EsBib extends EsBase {
...
 
-  parallelSeries_displayPacked () {
-    const displayPacked = this._parallelValueToIndexFromBasicMapping('series_displayPacked')
-    // Get parallel series without deduping
-    const parallelSeries = this._parallelValueToIndexFromBasicMapping('series', false)
-    if (!parallelSeries || !displayPacked) return null
-    const packed = parallelSeries.map((s, i) => (s && displayPacked[i]) ? `${s}||${displayPacked[i]}` : null).filter(Boolean)
-    return packed.length ? packed : null
+  parallelSeries () {
+    const values = this._parallelValueToIndexFromBasicMapping('series')
```
Comment (charmingduchess) [inline lib/es-models/bib.js]: I don't think it's a good idea to rely on mappings in two different repos (browse term vs bib-mapping.json) for series vs. series display components. I didn't think about that yesterday. If we can use browse-term module for display, why can't we use it for this property?

**Review by charmingduchess — CHANGES_REQUESTED** [review]

File: lib/es-models/bib.js (hunk 3 above)
Comment (7emansell) [reply lib/es-models/bib.js]: as discussed offline, bib mapping still exists but im using the browse term data now. Still using `extractDisplayComponents` for seriesUniformTitle and seriesAddedEntry simply bc it does the portioning/trailing v x removal i need to do anyways

File: lib/es-models/bib.js (hunk 1 above)
Comment (7emansell) [reply lib/es-models/bib.js]: updated and added unit tests to confirm this is still happening

File: lib/es-models/bib.js [hunk 4]
```diff
@@ -620,22 +618,18 @@ class EsBib extends EsBase {
     return this.browseTermData.contributorLiteral.map(term => term?.variant)
   }
 
-  parallelContributors_displayPacked () {
-    return this.browseTermData.contributorLiteral.map(term => {
-      if (!term || !term.variant) return null
-      return term.variant + '||' + term.variantLabel
-    }).filter(Boolean)
+  parallelContributors_displayComponents () {
```
Comment (charmingduchess) [inline lib/es-models/bib.js:628]: do parallels need to be filtered for redundant terms too?

File: lib/es-models/bib.js [hunk 5]
```diff
@@ -807,62 +785,95 @@ class EsBib extends EsBase {
...
-    // Get parallel series without deduping
-    const parallelSeries = this._parallelValueToIndexFromBasicMapping('series', false)
-    if (!parallelSeries || !displayPacked) return null
-    const packed = parallelSeries.map((s, i) => (s && displayPacked[i]) ? `${s}||${displayPacked[i]}` : null).filter(Boolean)
-    return packed.length ? packed : null
+  parallelSeries () {
+    const terms = this.browseTermData.series.filter(Boolean)
+    if (!terms.length) return null
+    // Maintain nulls for matching the primary array
+    let values = terms.map((term) => `${term.variant || ''}`.trim() || null)
```
Comment (charmingduchess) [inline lib/es-models/bib.js:846]: i think we're going to have to rethink how we're doing parallels. Right now they are not getting the same treatment as primary fields nor across parallels. We should make sure that the parallel display components and parallel non display fields have the same output for some edge cases or we're going potentially see bad link.

Comment (charmingduchess) [reply lib/es-models/bib.js:846]: The more we can perform the same transformations on these, ie have methods like `parallelDisplayComponent` and `parallelForNonDisplayComponentsThatNeedToCorrespondToDisplayComponent`, the easier it will be to test for these edge cases.

**Review by charmingduchess — COMMENTED** [review]
I'm a little worried about the heterogeneous nature of how the different fields are filtering (or not) redundant names and parallels. There should definitely be some spot checking in unit tests for the different properties when there are redundant names and parallels. We should also start planning to rethink how parallels are built and indexed...

File: lib/es-models/bib.js (hunk 5 above)
Comment (7emansell) [reply lib/es-models/bib.js:846]: added `parallelDisplayComponent` method

File: test/unit/es-bib.test.js [hunk 6]
```diff
@@ -857,16 +859,70 @@ describe('EsBib', function () {
...
       const record = new SierraBib(require('../fixtures/bib-23033611.json'))
       const esBib = new EsBib(record)
-      expect(esBib.parallelContributors_displayPacked()).to.deep.equal(
-        ['Народна библиотека "Стефан Првовенчани"||Народна библиотека "Стефан Првовенчани", issuing body']
+      expect(esBib.parallelContributors_displayComponents()).to.deep.equal(
+        [{ name: 'Народна библиотека "Стефан Првовенчани"', title: '', role: 'issuing body', label: 'Народна библиотека "Стефан Првовенчани", issuing body' }]
       )
     })
+
+    it('should filter out parallel contributor display components that duplicate parallel creators', () => {
```
Comment (7emansell) [inline test/unit/es-bib.test.js:907]: @charmingduchess new unit tests here

Comment (charmingduchess) [reply test/unit/es-bib.test.js:907]: These are great, and I know it's a drag and a half to get these tests set up. Can you make one more test that verifies that a display component and the non display property end up with the same indexes when there are not parallel values for every primary?
As in, a test that verifies a situation ending in: contributorLiteral_displayComponents array like ['contributor 1', 'contributor 2'], parallelContributorLiteral_displayComponents like [null, 'parallel contributor 2']

**Review by charmingduchess — CHANGES_REQUESTED** [review]

File: test/unit/es-bib.test.js [hunk 7]
```diff
@@ -857,16 +859,101 @@ describe('EsBib', function () {
...
+      const esBib = new EsBib(record)
+      expect(esBib.parallelCreators_displayComponents()).to.deep.equal([
+        { name: 'Параллельный создатель', title: '', role: '', label: 'Параллельный создатель' }
+      ])
+      expect(esBib.parallelContributors_displayComponents()).to.deep.equal([
+        { name: 'Параллельный участник', title: '', role: '', label: 'Параллельный участник' }
+      ])
+    })
+
+    it('parallelContributors_displayComponents and parallelContributorLiteral are index-aligned: null at positions where primary has no parallel', () => {
```
Comment (7emansell) [inline test/unit/es-bib.test.js:927]: null alignment here @charmingduchess

**Review by charmingduchess — APPROVED** [review]
heroic! thanks for getting the test coverage dialed in.
